feature: Add support for Generic OIDC authentication provider

s-starostin opened the pull request:
## Issue: 
Resolves [https://github.com/rancher/rancher/issues/51059](https://github.com/rancher/rancher/issues/51059)
 
## Problem
Currently, the Rancher2 Terraform provider does not support the configuration of a Generic OIDC authentication provider. This prevents users from automating the integration of OIDC-compliant identity providers (such as self-hosted GitLab, etc.) through Infrastructure as Code (IaC). All configuration for this provider must be done manually through the Rancher UI or direct API calls.

## Solution
This PR introduces a new resource, `rancher2_auth_config_generic_oidc`, to manage the Generic OIDC authentication provider in Rancher. The resource targets the `genericoidc` auth provider in the Rancher API.

The new resource allows users to:
- Enable and disable the Generic OIDC provider.
- Configure all necessary OIDC client options, including `client_id`, `client_secret`, and endpoint URLs.
- Specify the required `scopes` and the `groups_field` to correctly parse group memberships from the OIDC token.
- Explicitly enable group processing via the `group_search_enabled` flag, which is a crucial requirement for the `genericoidc` provider to correctly identify group principals.

This implementation allows for a complete, automated setup of Generic OIDC authentication, as demonstrated in the example below using GitLab as the identity provider.

Example Configuration:
```hcl
resource "rancher2_auth_config_generic_oidc" "gitlab" {
  enabled              = true
  client_id            = "your-gitlab-client-id"
  client_secret        = "your-gitlab-client-secret"
  issuer               = "https://gitlab.yourcompany.com"
  rancher_url          = "https://rancher.yourcompany.com/verify-auth"
  
  # Explicitly define endpoints for robustness
  auth_endpoint        = "https://gitlab.yourcompany.com/oauth/authorize"
  token_endpoint       = "https://gitlab.yourcompany.com/oauth/token"
  userinfo_endpoint    = "https://gitlab.yourcompany.com/oauth/userinfo"
  jwks_url             = "https://gitlab.yourcompany.com/oauth/discovery/keys"

  # Scopes required to get user and group info from GitLab
  scopes               = "openid profile email read_api"
  
  # OIDC claim for group membership
  groups_field         = "groups"

  # This flag is required to instruct Rancher to process group claims
  group_search_enabled = true
}

# Bind a GitLab group to the Rancher admin role
resource "rancher2_global_role_binding" "gitlab_admin_binding" {
  name               = "gitlab-admins"
  global_role_id     = "admin" # The name is case-sensitive
  group_principal_id = "genericoidc_group://rancher-admins"
  depends_on         = [rancher2_auth_config_generic_oidc.gitlab]
}
```
 
## Testing
The repro steps in the GitHub issue are valid as they describe the lack of the resource. Testing for this PR involves verifying the functionality of the new `rancher2_auth_config_generic_oidc` resource.

## Engineering Testing
### Manual Testing
Manually tested the full lifecycle (create, update, destroy) of the `rancher2_auth_config_generic_oidc` resource against a live Rancher v2.12 instance using a local Docker-based GitLab CE instance as the OIDC provider.

- **Create:** Successfully enabled and configured OIDC with GitLab. Verified that users from GitLab could log into Rancher.
- **Role Binding:** Successfully created a `rancher2_global_role_binding` to map a GitLab group to the Rancher "Admin" role. Verified that a user logging in as part of that group received the correct administrative permissions.
- **Update:** Successfully modified the `scopes` and `groups_field` of the OIDC configuration and confirmed the changes were applied in Rancher.
- **Destroy:** Successfully disabled the OIDC provider by running `terraform destroy`. Verified that the login method was removed from Rancher's login screen.

### Automated Testing
- **Unit Tests:** Added `structure_auth_config_generic_oidc_test.go` to provide comprehensive unit tests for the `flatten` and `expand` functions, ensuring correct mapping between the Terraform schema and the Rancher API model.
- **Acceptance Tests:** Added `resource_rancher2_auth_config_generic_oidc_test.go`, which includes a new acceptance test (`TestAccRancher2AuthConfigGenericOIDC_basic`). This test performs a full create-update-destroy lifecycle against a live Rancher API, verifying that all configurable fields are set and updated correctly.

## QA Testing Considerations
-   QA should test with a standard OIDC provider (like Keycloak or a public GitLab/Okta instance) to ensure compatibility.
-   **Fresh Install:** Test creating the `rancher2_auth_config_generic_oidc` resource from scratch. Verify that a user from the OIDC provider can log in. Test that a `rancher2_global_role_binding` correctly grants permissions to an OIDC group.
-   **Upgrade Scenario:** While this is a new resource, QA should ensure that applying this configuration does not negatively impact any other existing auth providers that might be configured.
-   **Key fields to test:**
    -   `enabled` (toggling between `true` and `false`).
    -   `group_search_enabled = true` is critical for group mapping to work.
    -   `access_mode = "restricted"` with `allowed_principal_ids` to ensure only specific users/groups can log in.
 
### Regressions Considerations
-   The risk of regression is very low. This change adds a new, self-contained resource and only modifies the `provider.go` file to register it. It does not alter any existing auth provider logic. The underlying Go client for Rancher is unchanged. The primary area to watch would be ensuring that the provider initialization and API client creation are not inadvertently affected, but the changes are minimal and well-scoped. Probability of regression is estimated to be less than 1%.

s-starostin committed:
feature: add rancher2_auth_config_generic_oidc

s-starostin commented:
Hi @matttrach,
I've just fixed the commit message to follow the conventional commit format required by the CI checks.
Сould you please approve a re-run?

Thanks for your help!

raulcabello commented on docs/resources/auth_config_generic_oidc.md:
```diff
@@ -0,0 +1,75 @@
+* `allowed_principal_ids` - (Optional) Allowed principal IDs for auth. Required if `access_mode` is `required` or `restricted`. Ex: `genericoidc_user://<USER_ID>` `genericoidc_group://<GROUP_ID>` (list)
+* `auth_endpoint` - (Optional/Computed) The OIDC Auth Endpoint URL.
+* `certificate` - (Optional/Sensitive) A PEM-encoded CA certificate for the OIDC provider.
+* `display_name_field` - (Optional/Computed) The name of the OIDC claim to use for the user's display name. Default `name` (string)
```
`display_name_field`  is not supported for the OIDC Generic Provider. Can you please remove it?

raulcabello commented on docs/resources/auth_config_generic_oidc.md:
```diff
@@ -0,0 +1,75 @@
+* `private_key` - (Optional/Sensitive) A PEM-encoded private key for the OIDC provider.
+* `scopes` - (Optional/Computed) The OIDC scopes to request. Defaults to `openid profile email` (string)
+* `token_endpoint` - (Optional/Computed) The OIDC Token Endpoint URL.
+* `uid_field` - (Optional/Computed) The name of the OIDC claim to use for the user's unique ID. Default `sub` (string)
```
same for `uid_field`

raulcabello commented on docs/resources/auth_config_generic_oidc.md:
```diff
@@ -0,0 +1,75 @@
+* `scopes` - (Optional/Computed) The OIDC scopes to request. Defaults to `openid profile email` (string)
+* `token_endpoint` - (Optional/Computed) The OIDC Token Endpoint URL.
+* `uid_field` - (Optional/Computed) The name of the OIDC claim to use for the user's unique ID. Default `sub` (string)
+* `user_name_field` - (Optional/Computed) The name of the OIDC claim to use for the user's username. Default `preferred_username` (string)
```
`user_name` is not supported for the Generic Provider at the moment. It's being implemented (see https://github.com/rancher/rancher/issues/49857). Can you please remove it?

raulcabello commented on rancher2/resource_rancher2_auth_config_generic_oidc_test.go:
```diff
@@ -0,0 +1,158 @@
+
+func TestAccRancher2AuthConfigGenericOIDC_basic(t *testing.T) {
+	if testAccRancher2AuthConfigGenericOIDCClientID == "" || testAccRancher2AuthConfigGenericOIDCClientSecret == "" || testAccRancher2AuthConfigGenericOIDCIssuerURL == "" {
+		t.Skip("Skipping test due to missing OIDC environment variables")
```
this test will always be skipped since we don't pass the env vars in CI. We don't have an env for testing this in CI and we don't have this kind of test for other providers, so I'd suggest to remove it.

raulcabello reviewed (changes requested):
@s-starostin Thanks a lot for the contribution! This PR looks good overall. I've left a few comments on some minor changes. Please take a look when you have a chance!

s-starostin committed:
feature: add rancher2_auth_config_generic_oidc

s-starostin commented:
@raulcabello , hi! 
All done. I've removed the unsupported fields from the docs and deleted the skipped test as requested.
Could you please take another look?

raulcabello reviewed (approved):
LGTM, thanks!

matttrach commented:
The build is failing: 
`rancher2/resource_rancher2_auth_config_generic_oidc_test.go:6:2: "testing" imported and not used`

s-starostin commented:
@matttrach , hello!
I've fixed this minor issue. Could you please check it again?

matttrach commented:
I have opened #1646 to track this PR's release.
